x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -404,15 +547,36 @@ SecretKey getDEKById(String dekId) throws IOException {
         }
 
         private SecretKey loadDEK(String dekId) throws IOException {
-            final BlobPath dekBlobPath = delegatedBasePath.add(DEK_ROOT_CONTAINER).add(dekId);
+            final SecureString repositoryPassword = repositoryPasswordSupplier.get();
+            // cache the generation that has been verified for the current password
+            // the generation is valid until the password changes and the new generation is not valid for the same password
+            final String passwordGeneration;
+            try {
+                passwordGeneration = this.loadGenerationForPasswordCache.computeIfAbsent(repositoryPassword, ignored -> {
+                    final Optional<String> inferredPasswordGeneration = inferLatestPasswordGeneration();
+                    if (inferredPasswordGeneration.isEmpty()) {
+                        throw new IllegalStateException("The repository is not initialized");
+                    }
+                    verifyPasswordForGeneration(repositoryPassword, inferredPasswordGeneration.get());
```

Comment:
This is (possibly) going to perform inline (not fork it) pbkdf2 computation, we discussed and it's OK.